File: skills/math-expression-parser/reference/src/parser.ts
/**
 * Parses a sequence of tokens into an AST using recursive descent.
 *
 * Operator precedence (lowest to highest):
 *   1. Addition, subtraction (+, -)
 *   2. Multiplication, division, modulo (*, /, %)
 *   3. Exponentiation (**)  — right-associative
 *   4. Unary minus (-)
 *   5. Atoms: numbers, parenthesized expressions
 *
 * @node parser
 * @depends-on token-types, ast-types
 * @contract parser.test.ts
 * @hint parser: Classic recursive descent. Each precedence level is a
 *       function that calls the next-higher level. Right-associativity
 *       for ** is handled by recursing into the same level instead of
 *       the next.
 */

import { type Token, type TokenKind } from "./token-types";
import {
  type AstNode,
  numberLiteral,
  unaryExpr,
  binaryExpr,
} from "./ast-types";

export function parse(tokens: Token[]): AstNode {
  let pos = 0;

  function peek(): Token | undefined {
    return tokens[pos];
  }

  function advance(): Token {
    const t = tokens[pos];
    pos++;
    return t;
  }

  function expect(kind: TokenKind): Token {
    const t = peek();
    if (!t || t.kind !== kind) {
      throw new Error(
        `Expected ${kind} but got ${t ? t.kind : "end of input"}`,
      );
    }
    return advance();
  }

  // Level 1: addition and subtraction (lowest precedence)
  function parseAddSub(): AstNode {
    let left = parseMulDiv();
    while (peek()?.kind === "plus" || peek()?.kind === "minus") {
      const opToken = advance();
      const op = opToken.kind === "plus" ? "+" : "-";
      const right = parseMulDiv();
      left = binaryExpr(op, left, right);
    }
    return left;
  }

  // Level 2: multiplication, division, modulo
  function parseMulDiv(): AstNode {
    let left = parsePower();
    while (
      peek()?.kind === "star" ||
      peek()?.kind === "slash" ||
      peek()?.kind === "percent"
    ) {
      const opToken = advance();
      const op =
        opToken.kind === "star" ? "*" : opToken.kind === "slash" ? "/" : "%";
      const right = parsePower();
      left = binaryExpr(op, left, right);
    }
    return left;
  }

  // Level 3: exponentiation (right-associative)
  function parsePower(): AstNode {
    const base = parseUnary();
    if (peek()?.kind === "power") {
      advance();
      const exponent = parsePower(); // right-recursive for right-associativity
      return binaryExpr("**", base, exponent);
    }
    return base;
  }

  // Level 4: unary minus
  function parseUnary(): AstNode {
    if (peek()?.kind === "minus") {
      advance();
      const operand = parseUnary(); // allow chained unary: --x
      return unaryExpr("-", operand);
    }
    return parseAtom();
  }

  // Level 5: atoms — numbers and parenthesized expressions
  function parseAtom(): AstNode {
    const t = peek();

    if (!t) {
      throw new Error("Unexpected end of input");
    }

    if (t.kind === "number") {
      advance();
      return numberLiteral(parseFloat(t.value));
    }

    if (t.kind === "lparen") {
      advance();
      const expr = parseAddSub();
      expect("rparen");
      return expr;
    }

    throw new Error(`Unexpected token: ${t.kind} '${t.value}'`);
  }

  const ast = parseAddSub();

  if (pos < tokens.length) {
    const remaining = tokens[pos];
    throw new Error(
      `Unexpected token after expression: ${remaining.kind} '${remaining.value}'`,
    );
  }

  return ast;
}

File: research/experiments/sweep/runs/mathexpr-full-csharp-sonnet45/MathExprParser/Parser.cs
namespace MathExprParser;

public static class Parser
{
    private class ParserState
    {
        public List<Token> Tokens { get; }
        public int Position { get; set; }

        public ParserState(List<Token> tokens)
        {
            Tokens = tokens;
            Position = 0;
        }

        public Token? Peek()
        {
            return Position < Tokens.Count ? Tokens[Position] : null;
        }

        public Token? Consume()
        {
            if (Position >= Tokens.Count) return null;
            return Tokens[Position++];
        }

        public bool IsAtEnd()
        {
            return Position >= Tokens.Count;
        }
    }

    public static AstNode Parse(List<Token> tokens)
    {
        var state = new ParserState(tokens);
        var result = ParseAddSub(state);

        if (!state.IsAtEnd())
        {
            throw new Exception("Unexpected token after expression");
        }

        return result;
    }

    private static AstNode ParseAddSub(ParserState state)
    {
        var left = ParseMulDiv(state);

        while (true)
        {
            var token = state.Peek();
            if (token == null) break;

            if (token.Kind == TokenKind.Plus || token.Kind == TokenKind.Minus)
            {
                state.Consume();
                var right = ParseMulDiv(state);
                left = new BinaryExpr(token.Value, left, right);
            }
            else
            {
                break;
            }
        }

        return left;
    }

    private static AstNode ParseMulDiv(ParserState state)
    {
        var left = ParsePower(state);

        while (true)
        {
            var token = state.Peek();
            if (token == null) break;

            if (token.Kind == TokenKind.Star || token.Kind == TokenKind.Slash || token.Kind == TokenKind.Percent)
            {
                state.Consume();
                var right = ParsePower(state);
                left = new BinaryExpr(token.Value, left, right);
            }
            else
            {
                break;
            }
        }

        return left;
    }

    private static AstNode ParsePower(ParserState state)
    {
        var left = ParseUnary(state);

        var token = state.Peek();
        if (token != null && token.Kind == TokenKind.Power)
        {
            state.Consume();
            var right = ParsePower(state); // Right-associative: recurse into same level
            return new BinaryExpr(token.Value, left, right);
        }

        return left;
    }

    private static AstNode ParseUnary(ParserState state)
    {
        var token = state.Peek();
        if (token != null && token.Kind == TokenKind.Minus)
        {
            state.Consume();
            var operand = ParseUnary(state); // Allow chained unary
            return new UnaryExpr("-", operand);
        }

        return ParseAtom(state);
    }

    private static AstNode ParseAtom(ParserState state)
    {
        var token = state.Consume();

        if (token == null)
        {
            throw new Exception("Unexpected end of input");
        }

        if (token.Kind == TokenKind.Number)
        {
            return new NumberLiteral(double.Parse(token.Value));
        }

        if (token.Kind == TokenKind.LParen)
        {
            var expr = ParseAddSub(state);
            var rparen = state.Consume();

            if (rparen == null || rparen.Kind != TokenKind.RParen)
            {
                throw new Exception("Expected rparen");
            }

            return expr;
        }

        throw new Exception($"Unexpected token: {token.Kind} '{token.Value}'");
    }
}
